cpp/serve/grammar/grammar_parser.cc

```diff
@@ -0,0 +1,350 @@
+  switch (Peek(-1)) {
+    case '*':
+      return bnf_grammar_->InsertRuleRef(HandleStarQuantifier(subrule_id));
+    case '+':
+      return bnf_grammar_->InsertRuleRef(HandlePlusQuantifier(subrule_id));
+    case '?':
+      return bnf_grammar_->InsertRuleRef(HandleQuestionQuantifier(subrule_id));
+    default:
+      LOG(FATAL) << "Unreachable";
+  }
+}
+
+BNFGrammarImpl::TSubruleId EBNFParser::ParseSequence() {
+  std::vector<BNFGrammarImpl::TSubruleId> elements;
+  elements.push_back(ParseQuantifier());
+  ConsumeSpace(in_parentheses_);
+  while (Peek() && Peek() != '|' && Peek() != ')' && Peek() != '\n' && Peek() != '\r') {
+    elements.push_back(ParseQuantifier());
+    ConsumeSpace(in_parentheses_);
+  }
+  return bnf_grammar_->InsertSequence(elements);
+}
+
+BNFGrammarImpl::TSubruleId EBNFParser::ParseOrRule() {
+  std::vector<BNFGrammarImpl::TSubruleId> choices;
+
+  choices.push_back(ParseSequence());
+  ConsumeSpace();
+  while (Peek() == '|') {
+    Consume();
+    ConsumeSpace();
+    choices.push_back(ParseSequence());
+    ConsumeSpace();
+  }
+  return bnf_grammar_->InsertOrRule(choices);
+}
+
+BNFGrammarImpl::Rule EBNFParser::ParseRule() {
+  std::string name = ParseName();
+  cur_rule_name_ = name;
+  ConsumeSpace();
+  if (Peek() != ':' || Peek(1) != ':' || Peek(2) != '=') {
+    ThrowParseError("Expect ::=");
+  }
+  Consume(3);
+  ConsumeSpace();
+  return {name, ParseOrRule()};
+}
+
+void EBNFParser::BuildRuleNameToId() {
+  ConsumeSpace();
+  while (Peek()) {
+    auto name = ParseName(true);
+    ConsumeSpace(false);
+    if (Peek() == ':' && Peek(1) == ':' && Peek(2) == '=') {
+      if (name.empty()) {
+        ThrowParseError("Expect rule name");
+      }
+      Consume(3);
+      if (rule_name_to_id_.count(name) != 0) {
+        ThrowParseError("Rule " + name + " is defined multiple times");
+      }
+      NewRule(name);
+    }
+    while (Peek() && Peek() != '\n' && Peek() != '\r') {
+      Consume();
+    }
+    ConsumeSpace();
+  }
+}
+
+void EBNFParser::Parse(String ebnf_string) {
+  Reset(ebnf_string.c_str());
+  BuildRuleNameToId();
+
+  Reset(ebnf_string.c_str());
+  ConsumeSpace();
+  while (Peek()) {
+    auto new_rule = ParseRule();
+    (*bnf_grammar_)[rule_name_to_id_[new_rule.name]].subrule = new_rule.subrule;
+
+    ConsumeSpace();
+  }
+
+  if (rule_name_to_id_.count("main") == 0) {
+    ThrowParseError("There must be a rule named main");
+  }
+}
+
+void EBNFParser::Reset(const char* cur) {
+  cur_ = cur;
+  cur_line_ = 1;
+  cur_column_ = 1;
+  cur_rule_name_ = "";
+  in_parentheses_ = false;
+}
+
+BNFGrammarImpl::TRuleId EBNFParser::NewRule(std::string name_hint) {
+  std::string name;
+  if (rule_name_to_id_.count(name_hint) == 0) {
+    name = name_hint;
+  } else {
+    int cnt = 1;
+    while (rule_name_to_id_.count(name_hint + "_" + std::to_string(cnt)) != 0) {
+      ++cnt;
+    }
+    name = name_hint + "_" + std::to_string(cnt);
+  }
+  auto rule_id = bnf_grammar_->InsertRule({name, -1});
```

Comment:
I am wondering what the `-1` means here. Will the value be updated later on? (and if so, where does it get updated?)

Reply:
Yeah, it is the default subrule. Later we will add a real subrule to it.